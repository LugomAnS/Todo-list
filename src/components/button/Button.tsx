import React from "react"

type ButtonPropsType = {
  name: string,
}

export function Button({name}: ButtonPropsType) {
  return (
    <button>{name}</button>
  )
}