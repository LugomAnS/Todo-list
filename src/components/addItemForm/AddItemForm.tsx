import { ChangeEvent, KeyboardEvent, memo, useState } from "react";
import { Button } from "../button/Button";

type AddItemFormPropsType = {
  onClick: (value: string) => void
}

function AddItemForm(props: AddItemFormPropsType) {
  const [taskTitle, setTaskTitle] = useState("");
  const [error, setError] = useState("");

  const callbacks = {
    onChangeTaskTitle: (e: ChangeEvent<HTMLInputElement>) => {
      error && setError("");
      setTaskTitle(e.currentTarget.value);
    },
    onAddTask: () => {
      const trimmedTitle = taskTitle.trim();

      if (!trimmedTitle) {
        setTaskTitle("");
        setError("Некорректное название");
        return;
      }
      props.onClick(trimmedTitle);
      setTaskTitle("");
    },
    onKeyDownAddTask: (e: KeyboardEvent<HTMLInputElement>) =>
    e.key === "Enter" && callbacks.onAddTask(),
  }

  return (
    <div>
      <input
        className={error ? "task-input-error" : ""}
        value={taskTitle}
        onChange={callbacks.onChangeTaskTitle}
        onKeyDown={callbacks.onKeyDownAddTask}
      />
      <Button name="+" onClick={callbacks.onAddTask} isDisabled={!taskTitle} />
      {error && <div style={{ color: "red" }}>{error}</div>}
    </div>
  );
}

export default memo(AddItemForm);
