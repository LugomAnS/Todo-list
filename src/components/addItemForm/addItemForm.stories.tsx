import type { Meta, StoryObj } from '@storybook/react';
import { fn } from '@storybook/test';

import AddItemForm, { AddItemFormPropsType } from './AddItemForm';
import { ChangeEvent, memo, useState, KeyboardEvent } from 'react';
import TextField from '@mui/material/TextField/TextField';
import Button from '@mui/material/Button';

// More on how to set up stories at: https://storybook.js.org/docs/writing-stories#default-export
const meta: Meta<typeof AddItemForm> = {
  title: 'ToDoList/AddItemForm',
  component: AddItemForm,
  parameters: {
    // Optional parameter to center the component in the Canvas. More info: https://storybook.js.org/docs/configure/story-layout
    layout: 'centered',
  },
  // This component will have an automatically generated Autodocs entry: https://storybook.js.org/docs/writing-docs/autodocs
  tags: ['autodocs'],
  // More on argTypes: https://storybook.js.org/docs/api/argtypes
  argTypes: {
    onClick: {
      description: "Button clicked inside form",
    },
  },
  // Use `fn` to spy on the onClick arg, which will appear in the actions panel once invoked: https://storybook.js.org/docs/essentials/actions#action-args
  args: {
    onClick: fn()
  },
};

export default meta;
type Story = StoryObj<typeof AddItemForm>;

// More on writing stories with args: https://storybook.js.org/docs/writing-stories/args
export const AddItemFormStory: Story = {};

const AddItemFormError = memo((props: AddItemFormPropsType) => {
  const [taskTitle, setTaskTitle] = useState("");
  const [error, setError] = useState("Некорректное название");

  const callbacks = {
    onChangeTaskTitle: (e: ChangeEvent<HTMLInputElement>) => {
      error && setError("");
      setTaskTitle(e.currentTarget.value);
    },
    onAddTask: () => {
      const trimmedTitle = taskTitle.trim();

      if (!trimmedTitle) {
        setTaskTitle("");
        setError("Некорректное название");
        return;
      }
      props.onClick(trimmedTitle);
      setTaskTitle("");
    },
    onKeyDownAddTask: (e: KeyboardEvent<HTMLInputElement>) =>
      e.key === "Enter" && callbacks.onAddTask(),
  };

  const styles = {
    maxWidth: "40px",
    maxHeight: "40px",
    minWidth: "40px",
    minHeight: "40px",
  };

  return (
    <div>
      <TextField id="outlined-basic"
               className={error ? "task-input-error" : ""}
               value={taskTitle}
               onChange={callbacks.onChangeTaskTitle}
               onKeyDown={callbacks.onKeyDownAddTask}
               size="small"
               placeholder="Введите значение"
               label={error ? error : "Введите значение"}
               error={!!error}
      />
      <Button
        sx={styles}
        variant="contained"
        onClick={callbacks.onAddTask}
        disabled={!taskTitle}
        size="small"
      >
        +
      </Button>
    </div>
  );
})


export const AddItemFormStoryError: Story = {
  render: () => <AddItemFormError onClick={() => {}} />
}