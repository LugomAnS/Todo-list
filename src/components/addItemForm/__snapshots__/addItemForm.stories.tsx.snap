// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ToDoList/AddItemForm AddItemFormStory smoke-test 1`] = `
<div>
  <div class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root">
    <label class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeSmall MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeSmall MuiInputLabel-outlined css-1pysi21-MuiFormLabel-root-MuiInputLabel-root"
           data-shrink="false"
           for="outlined-basic"
           id="outlined-basic-label"
    >
      Введите значение
    </label>
    <div class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-sizeSmall css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root">
      <input aria-invalid="false"
             id="outlined-basic"
             placeholder="Введите значение"
             type="text"
             class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputSizeSmall css-1n4twyu-MuiInputBase-input-MuiOutlinedInput-input"
             value
      >
      <fieldset aria-hidden="true"
                class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
      >
        <legend class="css-yjsfm1">
          <span>
            Введите значение
          </span>
        </legend>
      </fieldset>
    </div>
  </div>
  <button class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeSmall MuiButton-containedSizeSmall Mui-disabled MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeSmall MuiButton-containedSizeSmall css-1rzwrt7-MuiButtonBase-root-MuiButton-root"
          tabindex="-1"
          type="button"
          disabled
  >
    +
  </button>
</div>
`;

exports[`ToDoList/AddItemForm AddItemFormStoryError smoke-test 1`] = `
<div>
  <div class="MuiFormControl-root MuiTextField-root task-input-error css-1u3bzj6-MuiFormControl-root-MuiTextField-root">
    <label class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeSmall MuiInputLabel-outlined MuiFormLabel-colorPrimary Mui-error MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeSmall MuiInputLabel-outlined css-1pysi21-MuiFormLabel-root-MuiInputLabel-root"
           data-shrink="false"
           for="outlined-basic"
           id="outlined-basic-label"
    >
      Некорректное название
    </label>
    <div class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary Mui-error MuiInputBase-formControl MuiInputBase-sizeSmall css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root">
      <input aria-invalid="true"
             id="outlined-basic"
             placeholder="Введите значение"
             type="text"
             class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputSizeSmall css-1n4twyu-MuiInputBase-input-MuiOutlinedInput-input"
             value
      >
      <fieldset aria-hidden="true"
                class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
      >
        <legend class="css-yjsfm1">
          <span>
            Некорректное название
          </span>
        </legend>
      </fieldset>
    </div>
  </div>
  <button class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeSmall MuiButton-containedSizeSmall Mui-disabled MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeSmall MuiButton-containedSizeSmall css-1rzwrt7-MuiButtonBase-root-MuiButton-root"
          tabindex="-1"
          type="button"
          disabled
  >
    +
  </button>
</div>
`;
